crates/provider/src/layers/batch.rs

```diff
@@ -0,0 +1,515 @@
+            return false;
+        }
+        if params.overrides.as_ref().is_some_and(|overrides| !overrides.is_empty()) {
+            return false;
+        }
+        let tx = params.data();
+        if tx.to().is_none() {
+            return false;
+        }
+        macro_rules! any_is_some {
+            ($($name:ident),* $(,)?) => { false $(|| tx.$name().is_some())* };
+        }
+        // TODO: Not exhaustive, should we serialize to JSON and check that instead?
+        if any_is_some!(
+            chain_id,
+            nonce,
+            from,
+            value,
+            gas_price,
+            max_fee_per_gas,
+            max_priority_fee_per_gas,
+            gas_limit,
+            access_list
+        ) {
+            return false;
+        }
+        true
+    }
+
+    async fn schedule<N: Network>(
+        self,
+        msg: BatchProviderMessageKind<N>,
+    ) -> TransportResult<Bytes> {
+        let (msg, rx) = BatchProviderMessage::new(msg, self.m3a);
+        self.tx.send(msg).map_err(|_| TransportErrorKind::backend_gone())?;
+
+        let IMulticall3::Result { success, returnData: data } =
+            rx.await.map_err(|_| TransportErrorKind::backend_gone())??;
+        if !success {
+            let revert_data = if data.is_empty() { "" } else { &format!(" with data: {data}") };
+            return Err(TransportErrorKind::custom_str(&format!(
+                "multicall batched call reverted{revert_data}"
+            )));
+        }
+        Ok(data)
+    }
+
+    async fn schedule_and_decode<N: Network, T: SolValue>(
+        self,
+        msg: BatchProviderMessageKind<N>,
+    ) -> TransportResult<T>
+    where
+        T: From<<T::SolType as SolType>::RustType>,
+    {
+        let data = self.schedule(msg).await?;
+        T::abi_decode(&data, false).map_err(TransportErrorKind::custom)
+    }
+}
+
+struct BatchProviderBackend<P, N: Network = Ethereum> {
+    inner: Arc<P>,
+    m3a: Address,
+    wait: Duration,
+    rx: mpsc::UnboundedReceiver<BatchProviderMessage>,
+    pending: Vec<BatchProviderMessage>,
+    _pd: PhantomData<N>,
+}
+
+impl<P: Provider<N> + 'static, N: Network> BatchProviderBackend<P, N> {
+    fn spawn(inner: Arc<P>, layer: &BatchLayer) -> mpsc::UnboundedSender<BatchProviderMessage> {
+        let BatchLayer { m3a, wait } = *layer;
+        let (tx, rx) = mpsc::unbounded_channel();
+        let this = Self { inner, m3a, wait, rx, pending: Vec::new(), _pd: PhantomData };
+        this.run().spawn_task();
+        tx
+    }
+
+    async fn run(mut self) {
+        'outer: loop {
+            // Wait for the first message.
+            debug_assert!(self.pending.is_empty());
+            match self.rx.recv().await {
+                Some(msg) => self.process_msg(msg),
+                None => break,
+            }
+
+            // Handle all remaining messages after waiting the duration.
+            debug_assert!(!self.pending.is_empty());
+            sleep(self.wait).await;
```

Comment:
this overhead is now always applied per batch but it could be that the channel already has sufficient pending messages (but unlikely) so this is fine imo

Reply:
we'd still have to wait, if we had another limit then yes